[torch.compile] torch.vmap supports dynamic shapes + enable flex attention create_block_mask dynamic shapes

yanboliang opened the pull request:
Fixes #ISSUE_NUMBER


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

yanboliang committed:
[BE] Use unwrap_sym_or_const helper function

ghstack-source-id: b654199ee3762343f0c514c412aaa61b40917e66
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137077

yanboliang committed:
[torch.compile] torch.vmap supports dynamic shapes

ghstack-source-id: ecd76f89f3650b8d2edb8906ed611b1b6c65a5cf
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137075

yanboliang commented on test/functorch/test_vmap.py:
```diff
@@ -3366,6 +3366,7 @@ def f(x):
             )(x)
             self.assertEqual(output, expected)
 
+    @skipIfTorchDynamo()
```
All of these skips are because of dynamic shapes doesn't work well with eager backend, all passed if I switch to aot_eager or inductor backend.

zou3519 replied:
Why did this succeed before this PR? (I want to check that this PR isn't regressing any previous behavior)

yanboliang commented on torch/_dynamo/variables/lists.py:
```diff
@@ -88,12 +89,7 @@ def as_proxy(self):
-            index = arg.sym_num
-        else:
-            index = arg.as_python_constant()
+        index = unwrap_sym_or_const(arg)
```
Just BE work to reuse utility functions.

flynn-goldenowl commented:
Failed to generate code suggestions for PR

flynn-goldenowl commented:
Failed to generate code suggestions for PR

flynn-goldenowl commented:
## PR Code Suggestions ✨

<!-- 39d5be8 -->

<table><thead><tr><td>Category</td><td align=left>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </td></tr><tbody><tr><td rowspan=1><strong>Torch.compile optimisation</strong></td>
<td>



<details><summary>Use <code>vmap_batch_size</code> argument in <code>torch.compile</code> for optimization</summary>

___


**Consider using <code>torch.compile</code> with the <code>vmap_batch_size</code> argument to optimize <br>performance.**

[test/dynamo/test_dynamic_shapes.py [2432-2433]](https://github.com/pytorch/pytorch/pull/137163/files#diff-ccbb61cbbd4578f2d6958ad4bdcf34c9be5bed3ebe2affe8a1eee8d93724eee6R2432-R2433)

```diff
+@torch.compile(vmap_batch_size=4, backend="inductor")
 def fn(func, a, b, c, d):
   ...
   return func(a, b, c, d)
 
```                   
</details></td></tr><tr><td rowspan=1><strong>Replace_function_call</strong></td>
<td>



<details><summary>Replace <code>torch.jit.optimize_for_inference</code> with <code>_optimize_for_inference</code></summary>

___


**To fix the issue, you should use <code>torch.jit._optimize_for_inference</code> instead of <br><code>torch.jit.optimize_for_inference</code>.**

[src/file1.py [23]](https://github.com/pytorch/pytorch/pull/137163/files#diff-bf79e48362d5d7b16c86492e16a5750143926feb3e6f59917cf407b3c979e935R23-R23)

```diff
-torch.jit.optimize_for_inference(module)
+torch.jit._optimize_for_inference(module)
 
```                   
</details></td></tr><tr><td rowspan=1><strong>Optimize_model</strong></td>
<td>



<details><summary>Optimize for inference using <code>_optimize_for_inference</code></summary>

___


**Add `torch.jit._optimize_for_inference` call before compiling the model.**

[src/file2.py [45]](https://github.com/pytorch/pytorch/pull/137163/files#diff-dd38842f38b7464c64fa6dbba7ffd6161d6dba9756d460386ec8c05cab2fbc54R45-R45)

```diff
-module = torch.compile(model)
+module = torch.jit._optimize_for_inference(module, model)
 
```                   
</details></td></tr><tr><td rowspan=1><strong>Skip_torch_dynamo</strong></td>
<td>



<details><summary>Skip testing in Torch Dynamo</summary>

___


**Add `@skipIfTorchDynamo()` decorator to skip testing when using Torch Dynamo.**

[test/inductor/test_flex_attention.py [3496-3497]](https://github.com/pytorch/pytorch/pull/137163/files#diff-e3963412cc249e81fecfcf8774f5428b2b5e837ff3633ae13d8b7886ab5bc3b9R3496-R3497)

```diff
+@skipIfTorchDynamo()
 @parametrize("in_dim", [0, 1])
 
```                   
</details></td></tr><tr><td rowspan=3><strong>Code improvement</strong></td>
<td>



<details><summary>Simplify the <code>unwrap_sym_or_const</code> function</summary>

___


**The <code>unwrap_sym_or_const</code> function can be improved by removing unnecessary type <br>checks.**

[torch/_dynamo/utils.py [1375-1382]](https://github.com/pytorch/pytorch/pull/137163/files#diff-b99ae8eb082046d6353fe4ee536a0d846e1007c7d3b21aefae468ded18fbf13eR1375-R1382)

```diff
 def unwrap_sym_or_const(arg):
-    from .variables import SymNodeVariable
+    return arg.as_python_constant() if not isinstance(arg, SymNodeVariable) else arg.sym_num
 
-    if isinstance(arg, SymNodeVariable):
-        return arg.sym_num
-    else:
-        return arg.as_python_constant()
-
```                   
</details></td></tr><tr><td>



<details><summary>Improve the <code>getitem_const</code> method</summary>

___


**The <code>getitem_const</code> method can be improved by using the <code>unwrap_sym_or_const</code> function.**

[torch/_dynamo/variables/lists.py [92-95]](https://github.com/pytorch/pytorch/pull/137163/files#diff-dd806d3d1db1425299d6ef82e64dd799fcf8f4f97c9f4c4fbd4c5b64d67fb2dbR92-R95)

```diff
 def getitem_const(self, tx: "InstructionTranslator", arg: VariableTracker):
     index = unwrap_sym_or_const(arg)
+    return self.python_type()(self._as_proxy()[index])
 
-    if isinstance(index, slice):
-        # Set source to None because slicing a list gives a new local
-
```                   
</details></td></tr><tr><td>



<details><summary>Improve the <code>get_item_dyn</code> method</summary>

___


**The <code>get_item_dyn</code> method can be improved by using the <code>unwrap_sym_or_const</code> function.**

[torch/_dynamo/variables/lists.py [689-693]](https://github.com/pytorch/pytorch/pull/137163/files#diff-dd806d3d1db1425299d6ef82e64dd799fcf8f4f97c9f4c4fbd4c5b64d67fb2dbR689-R693)

```diff
 def get_item_dyn(self, tx: "InstructionTranslator", arg: VariableTracker):
     index = unwrap_sym_or_const(arg)
+    return self.python_type()(self._as_proxy()[index]) if isinstance(index, slice) else SizeVariable(self.items[index])
 
-    if isinstance(index, slice):
-        return SizeVariable(self.items[index])
-
```                   
</details></td></tr></tr></tbody></table>

flynn-goldenowl commented:
Failed to generate code suggestions for PR

Chillee commented on torch/_dynamo/variables/torch.py:
```diff
@@ -301,7 +302,7 @@ def call_function(
             return VmapIncrementNestingCtxManagerVariable.create(
                 tx,
-                [guard_if_dyn(x) for x in args],
+                [unwrap_sym_or_const(x) for x in args],
```
I don't think this is the right thing to call here. I suspect these should be proxies instead.

Chillee replied:
@ezyang Do you know offhand how this should be held? Context is that this is what should be put in the Dynamo FX graph, and one of these should allow for a SymInt. At this point, some of these may be SymIntNode objects.

zou3519 replied:
The batch_size eventually gets passed to https://github.com/pytorch/pytorch/blob/63bbf712d889de8e07ff4f4f1896b3af3817b100/torch/_dynamo/variables/ctx_manager.py#L438-L443.

For that you want to be able to pass it a proxy and you also want to be able to retrieve the actual SymInt (for the lines above it that directly call `_vmap_increment_nesting`.

I would pass batch_size as the SymNodeVariable here (so that the VmapIncrementNestingCtxManagerVariable holds VariableTrackers) and then do then unwrapping inside it.

NB: there's likely a pre-existing bug here, in that we shouldn't actually be directly calling `torch._C._vmap_increment_nesting` (instead we should be simulating it in Dynamo). But I don't think this matters here

Chillee commented on torch/nn/attention/flex_attention.py:
```diff
@@ -845,7 +845,7 @@ def causal_mask(b, h, q_idx, kv_idx):
     KV_LEN = _round_up_to_multiple(KV_LEN, KV_BLOCK_SIZE)
     if _compile:
-        inner_func = torch.compile(inner_func, fullgraph=True, dynamic=False)
+        inner_func = torch.compile(inner_func, fullgraph=True, dynamic=True)
```
Don't add `dynamic=True` - just remove the dynamic setting. Allow for automatic dynamic shapes I think.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -2415,6 +2415,35 @@ def fn(x):
+
+        a, b, c, d = 4, 8, 16, 16
+        self.assertEqual(fn(gn, a, b, c, d), compiled_fn(gn, a, b, c, d))
+        # Ensure no recompile if dynamic shapes enabled.
+        self.assertEqual(cnt.frame_count, ifdynstaticdefault(2, 1))
+
```
Let's do an expecttest test -- that'll show us if we're threading the SymNodeVariable's proxy through correctly.

zou3519 reviewed (changes requested):
I don't think we're threading the SymNodeVariable's proxies through correctly; let's add an expecttest on the dynamo graph to confirm

yanboliang committed:
Update

ghstack-source-id: 2055377daaab20ea501646c008509931bcf15f15
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137321

yanboliang committed:
update

ghstack-source-id: 48d1d4dd66c7111cbf68e10e2e8d1853f9c86bff
Pull Request resolved: https://github.com/pytorch/pytorch/pull/137322

yanboliang commented:
@pytorchbot merge

yanboliang commented:
@pytorchbot merge -f "non relevant failures"
